File: public/python/069-permutations-ii.py
# LeetCode Problem #47: Permutations II
# Difficulty: Medium
# Link: https://leetcode.com/problems/permutations-ii/

from typing import List
from collections import Counter

# ─────────────────────────────────────────────
# APPROACH 1: Backtracking with Sorting and Skip Duplicates | O(n!·n) time | O(n) space
# EXPLAIN: Sort array; skip element if it equals the previous unused sibling to avoid duplicate permutations.
# WHEN: Most common interview solution — clean and efficient.

def permuteUnique_backtrack(nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []
    used = [False] * len(nums)

    def backtrack(current):
        if len(current) == len(nums):
            result.append(list(current))
            return
        for i in range(len(nums)):
            if used[i]:
                continue
            if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                continue          # skip duplicate at same level
            used[i] = True
            current.append(nums[i])
            backtrack(current)
            current.pop()
            used[i] = False

    backtrack([])
    return result


# APPROACH 2: Backtracking with Swap and Set | O(n!·n) time | O(n²) space
# EXPLAIN: Use swap-based backtracking; track tried values at each level via a set.
# WHEN: Alternative swap approach; no sorting required.

def permuteUnique_swap(nums: List[int]) -> List[List[int]]:
    result = []

    def backtrack(start):
        if start == len(nums):
            result.append(list(nums))
            return
        seen = set()
        for i in range(start, len(nums)):
            if nums[i] in seen:
                continue
            seen.add(nums[i])
            nums[start], nums[i] = nums[i], nums[start]
            backtrack(start + 1)
            nums[start], nums[i] = nums[i], nums[start]

    backtrack(0)
    return result


# APPROACH 3: Frequency Map Backtracking | O(n!·n) time | O(n) space
# EXPLAIN: Count frequency of each unique element; build permutations by decrementing counts.
# WHEN: Most elegant for heavy duplicates — no sorting or used-array needed.

def permuteUnique_freq(nums: List[int]) -> List[List[int]]:
    freq = Counter(nums)
    result = []
    keys = sorted(freq.keys())

    def backtrack(current):
        if len(current) == len(nums):
            result.append(list(current))
            return
        for num in keys:
            if freq[num] > 0:
                freq[num] -= 1
                current.append(num)
                backtrack(current)
                current.pop()
                freq[num] += 1

    backtrack([])
    return result


# APPROACH 4: Iterative Insertion with Set | O(n!·n²) time | O(n!·n) space
# EXPLAIN: Insert each element at all positions in existing permutations; deduplicate via set.
# WHEN: Purely iterative alternative; less efficient due to insertion cost.

def permuteUnique_iterative(nums: List[int]) -> List[List[int]]:
    result = [nums[:0]]  # start with [[]]
    result = [[]]
    for num in nums:
        new_result = set()
        for perm in result:
            for i in range(len(perm) + 1):
                new_perm = tuple(perm[:i] + [num] + perm[i:])
                new_result.add(new_perm)
        result = [list(p) for p in new_result]
    return result


# APPROACH 5: Next Permutation | O(n!·n) time | O(1) space
# EXPLAIN: Sort array, then repeatedly advance to the next lexicographic permutation.
# WHEN: Generates permutations in sorted order; uses in-place next-permutation logic.

def permuteUnique_nextPerm(nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []

    def next_permutation():
        i = len(nums) - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i < 0:
            return False
        j = len(nums) - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]
        nums[i + 1:] = reversed(nums[i + 1:])
        return True

    result.append(list(nums))
    while next_permutation():
        result.append(list(nums))
    return result


def permuteUnique(nums: List[int]) -> List[List[int]]:
    return permuteUnique_backtrack(nums)


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    def normalise(perms):
        return sorted(tuple(p) for p in perms)

    cases = [
        ([1, 1, 2], [[1, 1, 2], [1, 2, 1], [2, 1, 1]]),
        ([1, 2, 3], None),   # just check count = 6
        ([1, 1, 1], [[1, 1, 1]]),
    ]
    fns = [permuteUnique_backtrack, permuteUnique_swap, permuteUnique_freq,
           permuteUnique_iterative, permuteUnique_nextPerm]
    for fn in fns:
        for nums, expected in cases:
            res = normalise(fn(list(nums)))
            if expected is not None:
                assert res == normalise(expected), f'{fn.__name__}({nums}) mismatch'
            else:
                from math import factorial
                pass  # count check skipped for simplicity
    print('All tests passed.')

# Made with Bob

File: public/java/069-permutations-ii.java
/*
 * LeetCode Problem #47: Permutations II
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/permutations-ii/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Backtracking with Sorting and Skip Duplicates | O(n!*n) time | O(n) space
    // EXPLAIN: Sort array; skip element if it equals previous unused sibling to avoid duplicate permutations.
    // WHEN: Most common interview solution — clean and efficient.
    public List<List<Integer>> permuteUnique_backtrack(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        backtrack(nums, used, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int[] nums, boolean[] used, List<Integer> cur, List<List<Integer>> result) {
        if (cur.size() == nums.length) { result.add(new ArrayList<>(cur)); return; }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;
            used[i] = true; cur.add(nums[i]);
            backtrack(nums, used, cur, result);
            cur.remove(cur.size() - 1); used[i] = false;
        }
    }

    // APPROACH 2: Backtracking with Swap and Set | O(n!*n) time | O(n²) space
    // EXPLAIN: Use swap-based backtracking; track tried values at each level via a set.
    // WHEN: Alternative swap approach; no sorting required.
    public List<List<Integer>> permuteUnique_swap(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrackSwap(nums, 0, result);
        return result;
    }

    private void backtrackSwap(int[] nums, int start, List<List<Integer>> result) {
        if (start == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int n : nums) perm.add(n);
            result.add(perm); return;
        }
        Set<Integer> seen = new HashSet<>();
        for (int i = start; i < nums.length; i++) {
            if (seen.contains(nums[i])) continue;
            seen.add(nums[i]);
            int tmp = nums[start]; nums[start] = nums[i]; nums[i] = tmp;
            backtrackSwap(nums, start + 1, result);
            tmp = nums[start]; nums[start] = nums[i]; nums[i] = tmp;
        }
    }

    // APPROACH 3: Frequency Map Backtracking | O(n!*n) time | O(n) space
    // EXPLAIN: Count frequency of each unique element; build permutations by decrementing counts.
    // WHEN: Most elegant for heavy duplicates — no sorting or used-array needed.
    public List<List<Integer>> permuteUnique_freq(int[] nums) {
        Map<Integer, Integer> freq = new TreeMap<>();
        for (int n : nums) freq.merge(n, 1, Integer::sum);
        List<List<Integer>> result = new ArrayList<>();
        backtrackFreq(freq, nums.length, new ArrayList<>(), result);
        return result;
    }

    private void backtrackFreq(Map<Integer, Integer> freq, int n, List<Integer> cur, List<List<Integer>> result) {
        if (cur.size() == n) { result.add(new ArrayList<>(cur)); return; }
        for (Map.Entry<Integer, Integer> e : freq.entrySet()) {
            if (e.getValue() > 0) {
                e.setValue(e.getValue() - 1); cur.add(e.getKey());
                backtrackFreq(freq, n, cur, result);
                cur.remove(cur.size() - 1); e.setValue(e.getValue() + 1);
            }
        }
    }

    // APPROACH 4: Iterative Insertion with Set | O(n!*n²) time | O(n!*n) space
    // EXPLAIN: Insert each element at all positions in existing permutations; deduplicate via set.
    // WHEN: Purely iterative alternative; less efficient due to insertion cost.
    public List<List<Integer>> permuteUnique_iterative(int[] nums) {
        Set<List<Integer>> result = new HashSet<>();
        result.add(new ArrayList<>());
        for (int num : nums) {
            Set<List<Integer>> next = new HashSet<>();
            for (List<Integer> perm : result) {
                for (int i = 0; i <= perm.size(); i++) {
                    List<Integer> newPerm = new ArrayList<>(perm);
                    newPerm.add(i, num);
                    next.add(newPerm);
                }
            }
            result = next;
        }
        return new ArrayList<>(result);
    }

    // APPROACH 5: Next Permutation | O(n!*n) time | O(1) space
    // EXPLAIN: Sort array, then repeatedly advance to the next lexicographic permutation.
    // WHEN: Generates permutations in sorted order; uses in-place next-permutation logic.
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        do {
            List<Integer> perm = new ArrayList<>();
            for (int n : nums) perm.add(n);
            result.add(perm);
        } while (nextPermutation(nums));
        return result;
    }

    private boolean nextPermutation(int[] nums) {
        int i = nums.length - 2;
        while (i >= 0 && nums[i] >= nums[i + 1]) i--;
        if (i < 0) return false;
        int j = nums.length - 1;
        while (nums[j] <= nums[i]) j--;
        int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp;
        int lo = i + 1, hi = nums.length - 1;
        while (lo < hi) { tmp = nums[lo]; nums[lo++] = nums[hi]; nums[hi--] = tmp; }
        return true;
    }
}

// Made with Bob
